python/kvikio/kvikio/zarr/_zarr_python_3.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+import asyncio
+import os
+from pathlib import Path
+
+import zarr.storage
+from zarr.abc.store import (
+    ByteRequest,
+    OffsetByteRequest,
+    RangeByteRequest,
+    SuffixByteRequest,
+)
+from zarr.core.buffer import Buffer, BufferPrototype
+from zarr.core.buffer.core import default_buffer_prototype
+
+import kvikio
+
+
+def _get(
+    path: Path, prototype: BufferPrototype, byte_range: ByteRequest | None
+) -> Buffer:
+    file_size = os.path.getsize(path)
+    file_offset: int
+
+    match byte_range:
+        case None:
+            nbytes = file_size
+            file_offset = 0
+        case OffsetByteRequest():
+            nbytes = max(0, file_size - byte_range.offset)
+            file_offset = byte_range.offset
+        case RangeByteRequest():
+            nbytes = byte_range.end - byte_range.start
+            file_offset = byte_range.start
+        case SuffixByteRequest():
+            nbytes = byte_range.suffix
+            file_offset = max(0, file_size - byte_range.suffix)
+        case _:
+            # This isn't allowed by mypy, but the tests assert we raise
+            # something here.
+            raise TypeError(f"Unexpected byte_range, got {byte_range}")
+
+    # kvikio doesn't support reading past the end of a file. Some zarr tests
+    # rely on this behavior: to "read" 3 bytes out of a 0 byte file, or to
+    # "seek" past the end of a file with file_offset. The semantics seem to
+    # be roughly the same as slicing an empty bytestring.
+
+    nbytes = min(nbytes, file_size)
+    file_offset = min(file_offset, file_size)
+
+    raw = prototype.nd_buffer.create(shape=(nbytes,), dtype="b").as_ndarray_like()
+    buf = prototype.buffer.from_array_like(raw)
+
+    with kvikio.CuFile(path) as f:
+        # Note: this currently creates an IOFuture and then blocks
+        # on reading it. The blocking read means this is in a regular
+        # sync function, and so this must be run in a threadpool.
+        future = f.pread(raw, size=nbytes, file_offset=file_offset)
+        future.get()  # blocks
+
+    return buf
+
+
+def _put(
+    path: Path,
+    value: Buffer,
+    start: int | None = None,
+    exclusive: bool = False,
+) -> int | None:
+    path.parent.mkdir(parents=True, exist_ok=True)
+    if start is not None:
+        with kvikio.CuFile(path, "r+b") as f:
+            f.write(value.as_array_like(), file_offset=start)
+        return None
+    else:
+        buf = value.as_array_like()
+        if exclusive:
+            if path.exists():
+                raise FileExistsError(f"File exists: {path}")
+        mode = "wb"
+        with kvikio.CuFile(path, flags=mode) as f:
+            return f.write(buf)
+
+
+class GDSStore(zarr.storage.LocalStore):
```

Comment:
Need a doctring

Reply:
These docstrings will be inherited from the [base class](https://zarr.readthedocs.io/en/stable/api/zarr/abc/store/index.html#zarr.abc.store.Store.get), which should suffice.